### PR TITLE
CI: set up a CI job for automated benchmarking (Linux, Python 3.9)

### PR DESCRIPTION
**Pull request by HarshCasper**
## Description

This PR:

- [X] Implements automated benchmarking on CI
- [X] Uploads ASV-generated results, HTML files and logs as artifacts
- [X] Adds build caching using `ccache` for faster builds
- [ ] Allows benchmark tests to be triggered using specific PR labels (Commented out as of now)

### CODE REVIEW

File: .github/workflows/benchmark.yml [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+    - name: Ccache performance
+      shell: bash -l {0}
+      run: ccache -s
+
+    - name: Run Benchmarks
+      id: benchmark
+      shell: bash -l {0}
+      env:
+        ASV_FACTOR: 1.5
```
Comment (rgommers) [inline .github/workflows/benchmark.yml:98]: This deserves a comment - I guess this means that if the performance decreases by 1.5x, a CI failure is triggered in the output parsing step?

**Review by rgommers — COMMENTED** [review]
This looks good @HarshCasper, thanks. A few high-level comments:
- The failure seems to be due to some issue in the logic; output says performance increased only.
- The runtime of the job was ~2.5 hrs, that is too long for a job triggered from a PR. We should seriously limited that, or run it 1x/night only. I thought the latter was the plan.
- Question on uploaded artifacts: how can we use them? Only download and inspect locally, or is there another plan for those?
- Does the "upload always" trigger a second failure if somehow the logs weren't generated? And is it a good idea to always upload if there was a failure earlier on (the answer here may depend on the usage question above)?

**Comment by rgommers** [comment]
> This leads to the CI failure. We can ideally remove/comment it out for now until the benchmarks are fixed on the upstream.

That sounds good to me. Can you make sure that's a separate commit though, so it's easy to get rid of? It will make merging things back into SciPy master more painful otherwise.

@tirthasheshpatel all of these benchmark failures are UNU.RAN related, maybe you could have a look? The `NumericalInverseHermite` ones are covered in https://github.com/scipy/scipy/issues/14667, the others are new here it looks like.

**Comment by rgommers** [comment]
The nightly run and `if: always` thoughts sound good @HarshCasper. 

> We need to download and utilize them further.

Does scikit-image do this at all? It can be left for a follow-up, it doesn't seem urgent. The scope of the current PR is fine I think.

**Comment by HarshCasper** [comment]
PR closed in favor of: https://github.com/rgommers/scipy/pull/129

**Pull request closed without merging**